pkg/identity/identity.go

```diff
@@ -0,0 +1,289 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package identity
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"io/ioutil"
+	"log"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/yahoo/athenz/libs/go/zmssvctoken"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/clusterconfig"
+	"github.com/yahoo/k8s-athenz-syncer/pkg/crypto"
+)
+
+var (
+	gracePeriod   = 5 * time.Minute
+	defaultExpiry = 24 * time.Hour
+)
+
+// PrivateKeyProvider returns a signing key on demand.
+type PrivateKeyProvider func() (*crypto.SigningKey, error)
+
+// tokenProvider provides tokens.
+type tokenProvider interface {
+	Token() (string, error)
+}
+
+type x509Provider interface {
+	X509Cert() (*tls.Config, error)
+}
+
+type X509Provider struct {
+	once         sync.Once
+	stop         chan struct{}
+	l            sync.RWMutex
+	current      *tls.Config
+	certFilePath string
+	keyFilePath  string
+	caCertFile   string
+	expiry       time.Time
+}
+
+// IdentityProvider provides ntokens and optional TLS certs.
+type IdentityProvider struct {
+	tokenProvider
+	x509Provider
+	domain  string
+	service string
+	podIP   net.IP
+	cc      *clusterconfig.ClusterConfiguration
+	ks      PrivateKeyProvider
+}
+
+// tp implements tokenProvider
+type tp struct {
+	domain  string
+	service string
+	ip      string
+	host    string
+	expiry  time.Duration
+	ks      PrivateKeyProvider
+	l       sync.Mutex
+	current string
+	expire  time.Time
+}
+
+// Config is the identity provider configuration.
+type Config struct {
+	Cluster                   *clusterconfig.ClusterConfiguration // cluster config
+	Domain                    string                              // Athenz domain
+	Service                   string                              // Athenz service
+	PodIP                     net.IP                              // Pod IP for IP SAN, nil ok for now
+	PrivateKeyProvider        PrivateKeyProvider                  // source for private keys
+	TokenHost                 string                              // ntoken hostname where available
+	TokenIP                   net.IP                              // ntoken IP address
+	TokenExpiry               time.Duration                       // token expire time
+	X509CertFile              string                              // file path for x509 cert
+	X509KeyFile               string                              // file path for x509 key
+	AthenzClientAuthnx509Mode bool                                // athenz client authentication. default is token.
+}
+
+// NewIdentityProvider returns an identity provider for the supplied configuration.
+// It returns either x509Provider or TokenProvider based on the presence of x509 configurations.
+func NewIdentityProvider(config Config) (*IdentityProvider, error) {
```

Comment:
We can port the identity provider without having a dependency on the cluster config and remove the X509 associated code as well since it won't be used.